### PR TITLE
Load imports recursively

### PR DESCRIPTION
**Pull request by EggBaconAndSpam**
This is the big change that enables us to implement 'semi-semantic' caching.

This change does not affect performance; however, we lose the ability to parameterise over `resolver` and `cacher`.

### CODE REVIEW

**Comment by Gabriel439** [comment]
I'm going to suggest that @AndrasKovacs reviews this since both of you are going to be working on similar parts of the implementation

**Comment by EggBaconAndSpam** [comment]
Sorry to leave this half-finished -- I didn't have time to look into the exception hierarchy. I will fix it on Monday.

**Comment by Gabriel439** [comment]
Also, I want to note that although change this removes support for `MonadCatch`-generalized utilities in `Dhall.Import`, they were originally added to solve a problem that no longer exists.  The original use case comes from this issue:

* https://github.com/dhall-lang/dhall-haskell/issues/92

... and this pull request:

* https://github.com/dhall-lang/dhall-haskell/pull/93

... which were trying to implement support for type synonyms using the import system.  Later, when the language added native support for type synonyms this feature became no longer necessary.

**Comment by EggBaconAndSpam** [comment]
> ... which were trying to implement support for type synonyms using the import system. Later, when the language added native support for type synonyms this feature became no longer necessary.

Thanks for the explanation!

**Comment by AndrasKovacs** [comment]
@Gabriel439 I can't say much about this PR besides what I said in #1129. At this point I don't know how much of this is reusable for me. A significant part of this PR in particular is probably not, code directly related to caching is more likely.

**Comment by EggBaconAndSpam** [comment]
@AndrasKovacs I don't expect much of this code to stay around. The actual caching stuff is going to be the next PR (and a much smaller and cleaner one at that).

File: dhall/src/Dhall/Import.hs [hunk 1]
```diff
@@ -486,86 +483,175 @@ chainImport (Chained parent) child@(Import importHashed@(ImportHashed _ (Remote
...
+                Right t -> return t
+            importSemantics <- case Dhall.Binary.decodeExpression term of
+                Left err -> throwMissingImport (Imported _stack err)
+                Right sem -> return sem
+            return (ImportSemantics {..})
 
+        Nothing -> do
+            ImportSemantics {..} <- loadImportFresh import_
+
+            let alphaNormal = Dhall.Core.alphaNormalize importSemantics
```
[the remark this answers is not in the document]
Comment (EggBaconAndSpam) [reply dhall/src/Dhall/Import.hs]: Well-caught! Turns out that `loadImportFresh` didn't alpha-normalise, thus violating that invariant. I fixed this, and moved the alpha normalisation step to `loadImportFresh` where it belongs!

File: dhall/src/Dhall/Import.hs [hunk 2]
```diff
@@ -486,86 +483,175 @@ chainImport (Chained parent) child@(Import importHashed@(ImportHashed _ (Remote
...
+        Just bytesStrict -> do
+            let actualHash = Crypto.Hash.hash bytesStrict
+            if hash == actualHash
+                then return ()
+                else do
+                    Status { _stack } <- State.get
+                    throwMissingImport (Imported _stack (HashMismatch {expectedHash = hash, ..}))
+
+            let bytesLazy = Data.ByteString.Lazy.fromStrict bytesStrict
+            -- ??
```
[the remark this answers is not in the document]
Comment (EggBaconAndSpam) [reply dhall/src/Dhall/Import.hs]: Ups. No, I didn't mean to leave that one xD

File: dhall/src/Dhall/Freeze.hs [hunk 3]
```diff
@@ -78,20 +77,12 @@ freezeImport directory _standardVersion import_ = do
...
+    let expressionHash = Dhall.Import.hashExpression _standardVersion normalizedExpression
 
-    let newImportHashed = (importHashed import_) { hash = expressionHash }
+    let newImportHashed = (importHashed import_) { hash = Just expressionHash }
 
     let newImport = import_ { importHashed = newImportHashed }
 
-    let rootImport :| _ = view stack status
-
-    State.evalStateT
```
[the remark this answers is not in the document]
Comment (EggBaconAndSpam) [reply dhall/src/Dhall/Freeze.hs]: Well caught! I underestimated the expectation that the frozen import would also be present in the cache! I'll fix this.

**Pull request merged**